Servers should always send CANCEL_PUSH (#3700)
quicwg/base-drafts: draft-ietf-quic-http.md

terminate the stream or MAY take no action.

When a server sends CANCEL_PUSH, it is indicating that it will not be fulfilling
a promise.  The client cannot expect the corresponding promise to be fulfilled.
A server SHOULD send a CANCEL_PUSH even if it has opened the corresponding

> stream.

Reviewer: Maybe not a direct result of this change but now I'm confused what happens if the server sends a CANCEL_PUSH that references a Push ID of a valid completed push stream. Do nothing seems like the right call.

Reviewer 2: I think that is the right call.

Reviewer 3: That matches what would occur in transport, so it does seem like the right call.

Author: Added text to this effect.  And that in fact could happen -- if we're suggesting a CANCEL_PUSH whenever the server abandons a stream (i.e. RESET_STREAM), it could be doing that to abort retransmissions even after writing the full payload.  The full payload might still have arrived if there's no loss.

Sending CANCEL_PUSH has no direct effect on the state of existing push streams.
A client SHOULD NOT send a CANCEL_PUSH when it has already received a

> corresponding push stream.  A push stream may arrive after a client has sent

Reviewer 4: ```suggestion
corresponding push stream.  A push stream could arrive after a client has sent
```

> CANCEL_PUSH, because a server may have not yet processed the CANCEL_PUSH. The

Reviewer 4: ```suggestion
CANCEL_PUSH, because a server might not have processed the CANCEL_PUSH. The
```
